File: lib/agent-kernel/call.ts
import { routeCompletion, type AgentRole as ModelRole } from '@/lib/model-router'
import { budgetExceeded } from './budget'
import type { Blackboard } from './blackboard'
import type { RunBudget } from './types'

/**
 * Enveloppe routeCompletion() avec la comptabilité de budget du kernel.
 * Refuse d'appeler le modèle si le budget est déjà épuisé — c'est ce qui
 * rend la boucle réflexion → décision "bornée et prévisible" plutôt que de
 * compter uniquement sur un nombre d'itérations fixe.
 */
export async function callAgentModel(
  role: ModelRole,
  system: string,
  userMessage: string,
  maxTokens: number,
  blackboard: Blackboard,
  budget: RunBudget
): Promise<{ text: string; model_used: string }> {
  const exceeded = budgetExceeded(budget, blackboard)
  if (exceeded) {
    return { text: '', model_used: `skipped (${exceeded})` }
  }

  blackboard.recordModelCall()
  return routeCompletion(role, system, userMessage, maxTokens)
}

File: backend/agent_kernel/call.py
"""Port de lib/agent-kernel/call.ts.

Enveloppe route_completion() avec la comptabilité de budget du kernel.
Refuse d'appeler le modèle si le budget est déjà épuisé.
"""
from model_router import route_completion

from .blackboard import Blackboard
from .budget import budget_exceeded
from .types import RunBudget


async def call_agent_model(
    role: str,
    system: str,
    user_message: str,
    max_tokens: int,
    blackboard: Blackboard,
    budget: RunBudget,
) -> dict[str, str]:
    exceeded = budget_exceeded(budget, blackboard)
    if exceeded:
        return {"text": "", "model_used": f"skipped ({exceeded})"}

    blackboard.record_model_call()
    return await route_completion(role, system, user_message, max_tokens)
